Cleanup README.md

jpomykala opened the pull request:
I rearranged a lot of content, making it a bit more modern. I focused on showing new users HOW TO USE it instead feeding them with the FabricJS history 😉

# Changes:
- Removed not working links
- Made a sponsorship section
- Removed duplicated links to resources
- Created a README header with the most important links
- Added a Quick Start section
- Added ReactJS example code
- Removed Bower installation instruction (I left a badge)
- Simplified sample HTML code

# EDIT

I didn't want to delete too much, because I'm fairly new in FabricJS. I would love to keep the docs and README.md page clean and simple. My intention was to remove duplicated and old/useless content, it's easy to get lost in all those resources and links.

ShaMan123 commented on README.md:
```diff
@@ -1,78 +1,50 @@
-<h3 id="bower-install">Install with bower</h3>
 
-    $ bower install fabric
+Using Fabric.js, you can create and populate objects on canvas; objects like simple geometrical shapes — rectangles, circles, ellipses, polygons, or more complex shapes consisting of hundreds or thousands of simple paths. You can then scale, move, and rotate these objects with the mouse; modify their properties — color, transparency, z-index, etc. You can also manipulate these objects altogether  ...
```
I think this should be written in bullets

ShaMan123 replied:
And what about adding some gifs?

ShaMan123 commented on README.md:
```diff
@@ -1,78 +1,50 @@
-Please follow the instructions located [here](https://github.com/Automattic/node-canvas#compiling) in order to get it up and running.
+<a href="http://fabricjs.com/fabric-intro-part-1">Introduction</a>
+<span>&nbsp;&nbsp;•&nbsp;&nbsp;</span>
+<a href="http://fabricjs.com/docs/">Docs</a>
```
I don't think this part is needed, github parses the page and creates a navbar

jpomykala replied:
[https://github.com/reactchartjs/react-chartjs-2](https://github.com/reactchartjs/react-chartjs-2) was my "inspiration" here. 😄 I didn't know that Github creates navbar somewhere. Where I can find it?

ShaMan123 replied:
In the README page.
![צילום מסך 2022-05-17 233113](https://user-images.githubusercontent.com/34343793/168904786-a4a69179-d0c3-4ce3-8b88-9009e625032e.png)

ShaMan123 commented on README.md:
````diff
@@ -86,19 +58,73 @@ Or you can use this instead if your build pipeline supports ES6 imports:
 import { fabric } from "fabric";
 ```
 
+NOTE: If you are using Fabric.js in a Node.js script, you will depend from [node-canvas](https://github.com/Automattic/node-canvas).`node-canvas` is an HTML canvas replacement that works on top of native libraries. Please [follow the instructions](https://github.com/Automattic/node-canvas#compiling) to get it up and running.
````
```suggestion
NOTE: If you are using Fabric.js in a Node.js script, you will depend on [node-canvas](https://github.com/Automattic/node-canvas). `node-canvas` is an HTML canvas replacement that works on top of native libraries. Please [follow the instructions](https://github.com/Automattic/node-canvas#compiling) to get it up and running.
```

I wonder if the instructions are up to date.

jpomykala replied:
My suggestion is to remove it completely and leave the link to it somewhere.

ShaMan123 commented on README.md:
```diff
@@ -86,19 +58,73 @@ Or you can use this instead if your build pipeline supports ES6 imports:
 
+NOTE: If you are using Fabric.js in a Node.js script, you will depend from [node-canvas](https://github.com/Automattic/node-canvas).`node-canvas` is an HTML canvas replacement that works on top of native libraries. Please [follow the instructions](https://github.com/Automattic/node-canvas#compiling) to get it up and running.
+
 NOTE: es6 imports won't work in browser or with bundlers which expect es6 module like vite. Use commonjs syntax instead.
```
Hmm this will soon be wrong THANKFULLY

ShaMan123 commented on README.md:
````diff
@@ -86,19 +58,73 @@ Or you can use this instead if your build pipeline supports ES6 imports:
+</script>
+```
+
+#### ReactJS
````
This is great

ShaMan123 commented on README.md:
```diff
@@ -86,19 +58,73 @@ Or you can use this instead if your build pipeline supports ES6 imports:
-<h3 id="fabric-building">Building</h3>
 
-1. [Install Node.js](https://github.com/joyent/node/wiki/Installation)
+<h3 id="fabric-building">Building</h3>
```
I vote to get rid of this entire section for 3 main reasons:
1. too esoteric
2. there's a new cli for fabric that is very simple to use - PRs should be made to make it better instead of this bloated part about building options.
3. The old build script is at it's end of life and will soon die out.

jpomykala replied:
I also vote for removing this part or move it to another document and just leave the link to it somewhere.

ShaMan123 commented on README.md:
```diff
@@ -115,74 +141,60 @@ See [the example section](#examples-of-use) for usage examples.
 
         $ npm run lint && npm run lint_tests
 
 <h3 id="fabric-building">Testing</h3>
```
Should this (and all the rest) be:
```suggestion
# Testing
```

jpomykala replied:
This is also a good idea. I left all those HTML tags with unchanged IDs because I wasn't sure if somebody didn't link to this paragraph somewhere else (fabricjs.com?)

ShaMan123 replied:
I don't think there are...
Bad practice IMO.
I think it's safe to remove

ShaMan123 commented on README.md:
````diff
@@ -115,74 +141,60 @@ See [the example section](#examples-of-use) for usage examples.
+```
 
-Make sure testem is installed
+#### 2. Install testem
````
not required, it's installed in devs

ShaMan123 commented on README.md:
```diff
@@ -212,66 +224,63 @@ For example:
 
-Get help in Fabric's IRC channel — irc://irc.freenode.net/#fabric.js
 
 ### Credits
```
This is old and outdated as well
I think it's irrelevant in the README - github does it why better.

ShaMan123 commented on README.md:
```diff
@@ -212,66 +224,63 @@ For example:
+- [Fabric.js on jsfiddle](https://jsfiddle.net/user/fabricjs/fiddles/)
+- [Fabric.js on Codepen.io](https://codepen.io/tag/fabricjs)
+- [Presentation from BK.js](http://www.slideshare.net/kangax/fabricjs-building-acanvaslibrarybk)
+- [Presentation from Falsy Values](http://www.slideshare.net/kangax/fabric-falsy-values-8067834)
```
These 2 last links are really old...

asturur replied:
Bountysource too i deleted the account after they changed the policies and eat up all the bounties

jpomykala replied:
Done

ShaMan123 commented on README.md:
```diff
@@ -115,74 +141,60 @@ See [the example section](#examples-of-use) for usage examples.
-3. Install NPM packages
-
-        $ npm install
+#### 1. Install NPM packages
```
Well the testing instructions are bad.
I think it's better to refer to CONTRIBUTING.md

ShaMan123 reviewed (commented):
It's been waiting for someone to pick it up.
Good job.
Added my comments.
I guess the rest of the team will have different input and ideas.
Are you aware that the website is due for a major version? I saw you contributed to it in the past. It could use your help.

jpomykala commented:
> It's been waiting for someone to pick it up.
> Good job.
> Added my comments.
> I guess the rest of the team will have different input and ideas.
> Are you aware that the website is due for a major version? I saw you contributed to it in the past. It could use your help.

I didn't know that but I would love to help with it! I tried to improve the documentation website in past but I didn't get how it works and why there is so much duplicated content. 

I've some experience with building documentation pages from scratch. 😉

ShaMan123 commented:
> I didn't know that but I would love to help with it! I tried to improve the documentation website in past but I didn't get how it works and why there is so much duplicated content.

Ohh man for sure. I can help. I only know the way in because @asturur showed it to me. Or else still I wouldn't dare come close to the website docs. 

> I've some experience with building documentation pages from scratch. 😉

Great! Help is much appreciated and the website is in dire need.

asturur commented:
I m fine with a refresh of the readme,
i think from the history parts we should keep the part where we say that is heavily inspired from, and add it on the credits as a bullet item.

I would also make the react example a bit more generic,

```tsx
import React, {useEffect, useRef} from 'react'
import {fabric} from 'fabric'
const FabricJSCanvas = () => {
  const canvasEl = useRef(null)
  useEffect(() => {
    const options = { ... };
    const canvas = new fabric.Canvas(canvasEl.current, options);
    updateCanvasContext(canvas);
    return () => {
      updateCanvasContext(null);
      canvas.dispose()
    }
  }, []);
  
  return (<canvas width="300" height="300" ref={canvasEl}/>)
});
export default FabricJSCanvas;
```

Something that let you understand that yes you initialize the canvas in this way, but then if you want to do something with it you have to expose the fabric.Canvas to the app in some way.

jpomykala commented:
I applied changes from comments.

@asturur I saw also you were trying to use Docusaurus for docs. Why it didn't work for FabricJS? I'd go with some popular solution rather than implementing docs from scratch with Gatsby. I think it's easier for people to do changes and it's much easier to maintain.

asturur commented:
I wasn't able to theme docusaurus as i wanted, i gave up quickly. Then someone came up with this gatsby prototype and i let him run.
As long as we can write MD files i m fine with any solution.

ShaMan123 commented on README.md:
```diff
@@ -225,7 +216,7 @@ For example:
 
 
-### Goals
+## Goals
```
I think this is more `Guidelines` than `Goals` because they are acheived.

jpomykala replied:
Achievements? 😄  Section contains great stuff, but I don't know how to name it properly.

ShaMan123 commented:
I think history belongs to an **About** section in the website.

asturur commented:
Yes History can be removed, we should just credit that demo developer in the credits.

asturur commented:
In the credits are missing @ShaMan123  and @melchiar

melchiar commented on README.md:
```diff
@@ -212,67 +215,65 @@ For example:
+
+## History
+
+Fabric.js started as a foundation for design editor on [printio.ru](http://printio.ru) — interactive online store with ability to create your own designs. The idea was to create [Javascript-based editor](http://printio.ru/ringer_man_tees/new), which would make it easy to manipulate vector shapes and images on T-Shirts. Since performance was one of the most critical requirements, we chose canvas o ...
```
```suggestion
Fabric.js started as a foundation for design editor on [printio.ru](http://printio.ru) — interactive online store with ability to create your own designs. The idea was to create [Javascript-based editor](http://printio.ru/ringer_man_tees/new), which would make it easy to manipulate vector shapes and images on T-Shirts. Since performance was one of the most critical requirements, we chose canvas over SVG. While SVG is excellent with static shapes, it's not as performant as canvas when it comes to dynamic manipulation of objects (movement, scaling, rotation, etc.). Fabric.js was heavily inspired by a canvas experiment created by Ernest Delgado. In fact, code from Ernest's experiment was the foundation of an entire framework. Later, Fabric.js grew into a collection of distinct object types and got an SVG-to-canvas parser.
```

This canvas experiment URL is actually long dead and I can find no new location for it so probably best to just remove the link.

melchiar replied:
actually nevermind, @asturur is right this whole section is kind of unnecessary

> Yes History can be removed, we should just credit that demo developer in the credits.

melchiar commented:
This is great, thanks @jpomykala for the submission!

jpomykala commented:
I added @melchiar and @ShaMan123. I used links to GitHub profiles, feel free to change it or let me know if I should change something there. 😄 I also removed History section, demo author was already in the Credits. 

@asturur Great, thanks! I will look at Docusaurus then. From my experience Gatsby is easy to break due to plugins compatibility and it's bit hard to debug due to GraphQL. I think the biggest challenge will be moving already written documentation. Let me know if you guys have any requirements regarding the documentation. Requirements so far:
- markdown 
- deployable on GitHub pages
- simple usage & clean UI
- working demos
- working FabricJS builder

ShaMan123 commented:
> working FabricJS builder

I'm not sure this needed for the next version.

Requirements:

- tests: We have tests running in the website. Perhaps it's time to think if we want tests to be part of the website or have another repo dedicated for browser tests (IMO the latter).  Another option is to use [puppeteer](https://github.com/puppeteer/puppeteer) for chrome and find something for firefox...
Regardless, we will want to keep the stress tests in tact so perhaps it's pointless to drop website testing.

- If we're up for it - I think the main page should be a crazy showcase of capabilities - instead of having a **demos** section so a dev lands there and can understand/feel if fabric is for them - and get blown away. This means creating an app like kitchensink but with design and UX

asturur commented:
> ink the biggest challenge will be moving already written documentation. Let me know if you guys have any requirements regarding the documentation

If you have an idea for docusaurus:

- easy to add docs
- nicely themeable as a modern website you see around on js libs
- chance to add custom code ( we need to support the small ad banner at least)

asturur commented:
oh must be fruible from the mobile phone. Usable and readable.
I don't think code demos should look perfect from mobile phone.

Jus that.
The homepage is a matter on it own, could be one of those long scrollable that shows you 3-4-5 things, a nice banner, an intro, a link to examples and docs, contributors and some sponsor maybe. Nothing extreme

Important is that it load fast. it does not have to have a canvas at all costs. It could load one after scrolling maybe
I have really no strong desire, a working docosaurus setup with a nice theme is fine enough.

asturur commented:
If we need to refine the readme even more we can open another PR, don't want to drag simple things too long.
